pyfr/integrators/dual/pseudo/pseudocontrollers.py

```diff
@@ -90,3 +97,120 @@ def pseudo_advance(self, tcurr, tout, dt):
 
         # Update
         self.finalise_pseudo_advance(self._idxcurr)
+
+
+class DualPseudoPIController(BaseDualPseudoController):
+    pseudo_controller_name = 'local-pi'
+
+    def __init__(self, *args, **kwargs):
+        super().__init__(*args, **kwargs)
+        sect = 'solver-time-integrator'
+
+        # Error tolerance
+        atol = self.cfg.getfloat(sect, 'atol')
+
+        # Error norm
+        self._norm = self.cfg.get(sect, 'errest-norm', 'l2')
+        if self._norm not in {'l2', 'uniform'}:
+            raise ValueError('Invalid error norm')
+
+        # PI control values
+        alpha = self.cfg.getfloat(sect, 'pi-alpha', 0.7)
+        beta = self.cfg.getfloat(sect, 'pi-beta', 0.4)
+
+        # Constants
+        maxf = self.cfg.getfloat(sect, 'max-fact', 1.01)
+        minf = self.cfg.getfloat(sect, 'min-fact', 0.98)
+        saff = self.cfg.getfloat(sect, 'safety-fact', 0.8)
+        dtau_maxf = self.cfg.getfloat(sect, 'dtau-max-fact', 3.0)
+
+        sord = self._pseudo_stepper_order
+        expa = alpha / sord
+        expb = beta / sord
+
+        # Limits for the local pseudo-time-step size
+        dtau_min = self._dtau
+        dtau_max = dtau_maxf * dtau_min
+
+        # Template parameters
+        tplargs = dict(ndims=self.system.ndims, nvars=self.system.nvars,
```

Comment:
Since these coefficients do not seem to be used elsewhere it may be cleaner do say:

tplargs = {'ndims': self.system.ndims, 'nvars': self.system.nvars}
tplargs['alpha'] = self.cfg.getfloat(sect, 'pi-alpha', 0.7)
...